File: apps/rust/src/auth/service.rs
use chrono::{Duration, Utc};
use redis::aio::ConnectionManager;
use uuid::Uuid;

use crate::db::UserRow;

use super::repository;
use super::session::{LoginRequest, SessionConfig, SharedSession};

pub struct EnsureSessionResult {
    pub session: SharedSession,
    pub created: bool,
}

pub async fn ensure_session(
    conn: &mut ConnectionManager,
    config: &SessionConfig,
    client_session_id: Option<&str>,
    request_session: Option<&SharedSession>,
) -> Result<EnsureSessionResult, redis::RedisError> {
    let now = Utc::now();
    if let Some(session) = request_session {
        if !session.is_expired(now) {
            return Ok(EnsureSessionResult {
                session: session.clone(),
                created: false,
            });
        }
    }
    if let Some(id) = client_session_id.map(str::trim).filter(|s| !s.is_empty()) {
        if let Some(session) = repository::find_by_id(conn, config, id).await? {
            if !session.is_expired(now) {
                return Ok(EnsureSessionResult {
                    session,
                    created: false,
                });
            }
            let _ = repository::delete(conn, config, id).await;
        }
    }
    let session = create_anonymous_session(conn, config).await?;
    Ok(EnsureSessionResult {
        session,
        created: true,
    })
}

pub async fn login(
    conn: &mut ConnectionManager,
    config: &SessionConfig,
    pool: &sqlx::PgPool,
    body: &LoginRequest,
) -> Result<SharedSession, AuthServiceError> {
    let user = resolve_user(pool, body).await?;
    let issued_at = Utc::now();
    let expires_at = issued_at + Duration::seconds(config.ttl_secs as i64);
    let session = SharedSession {
        session_id: Uuid::new_v4().to_string(),
        user_id: user.id,
        email: Some(user.email),
        name: user.name,
        issued_at,
        expires_at,
        issuer: "rust".into(),
    };
    repository::save(conn, config, &session)
        .await
        .map_err(AuthServiceError::Redis)?;
    Ok(session)
}

pub async fn logout(
    conn: &mut ConnectionManager,
    config: &SessionConfig,
    session_id: &str,
) -> Result<(), redis::RedisError> {
    repository::delete(conn, config, session_id).await
}

async fn create_anonymous_session(
    conn: &mut ConnectionManager,
    config: &SessionConfig,
) -> Result<SharedSession, redis::RedisError> {
    let issued_at = Utc::now();
    let expires_at = issued_at + Duration::seconds(config.ttl_secs as i64);
    let session = SharedSession {
        session_id: Uuid::new_v4().to_string(),
        user_id: 0,
        email: None,
        name: "Guest".into(),
        issued_at,
        expires_at,
        issuer: "rust".into(),
    };
    repository::save(conn, config, &session).await?;
    Ok(session)
}

async fn resolve_user(pool: &sqlx::PgPool, body: &LoginRequest) -> Result<UserRow, AuthServiceError> {
    if let Some(email) = body.email.as_deref().map(str::trim).filter(|s| !s.is_empty()) {
        return crate::db::find_user_by_email(pool, email)
            .await
            .map_err(AuthServiceError::Db)?
            .ok_or_else(|| AuthServiceError::NotFound(format!("No user with email {email}")));
    }
    if let Some(user_id) = body.user_id {
        return crate::db::find_user_by_id(pool, user_id)
            .await
            .map_err(AuthServiceError::Db)?
            .ok_or_else(|| AuthServiceError::NotFound(format!("No user with id {user_id}")));
    }
    Err(AuthServiceError::BadRequest(
        "email or userId is required".into(),
    ))
}

#[derive(Debug)]
pub enum AuthServiceError {
    BadRequest(String),
    NotFound(String),
    Db(sqlx::Error),
    Redis(redis::RedisError),
}

File: apps/zig/src/auth/service.zig
const std = @import("std");
const db_mod = @import("../db.zig");
const redis_mod = @import("../redis_client.zig");
const password_mod = @import("password.zig");
const repository = @import("repository.zig");
const session_mod = @import("session.zig");

pub const AuthError = error{
    RedisUnavailable,
    BadRequest,
    NotFound,
    Unauthorized,
    DatabaseUnavailable,
    OutOfMemory,
};

pub const EnsureResult = struct {
    session: session_mod.SharedSession,
    created: bool,
};

pub fn ensureSession(
    client: *redis_mod.Client,
    config: *const session_mod.SessionConfig,
    allocator: std.mem.Allocator,
    client_session_id: ?[]const u8,
) AuthError!EnsureResult {
    if (client_session_id) |id| {
        const trimmed = std.mem.trim(u8, id, " \t\r\n");
        if (trimmed.len > 0) {
            const found = repository.findById(client, config, allocator, trimmed) catch return error.RedisUnavailable;
            if (found) |session| {
                if (!session.isExpired()) {
                    return .{ .session = session, .created = false };
                }
                var expired = session;
                defer expired.deinit(allocator);
                repository.deleteSession(client, config, allocator, trimmed) catch {};
            }
        }
    }
    var session = createGuestSession(allocator) catch return error.OutOfMemory;
    errdefer session.deinit(allocator);
    repository.save(client, config, allocator, &session) catch return error.RedisUnavailable;
    return .{ .session = session, .created = true };
}

pub fn login(
    client: *redis_mod.Client,
    config: *const session_mod.SessionConfig,
    allocator: std.mem.Allocator,
    database: *db_mod.Db,
    email: []const u8,
    password: []const u8,
) AuthError!session_mod.SharedSession {
    const trimmed_email = std.mem.trim(u8, email, " \t\r\n");
    if (trimmed_email.len == 0) return error.BadRequest;
    const auth_row = database.findUserAuthByEmail(allocator, trimmed_email) catch return error.DatabaseUnavailable;
    const row = auth_row orelse return error.NotFound;
    defer db_mod.Db.freeUserAuthRow(row, allocator);
    const hash = row.password_hash orelse return error.Unauthorized;
    if (!password_mod.verifyPassword(password, hash)) return error.Unauthorized;
    var session = createUserSession(allocator, row.id, row.name, row.email) catch return error.OutOfMemory;
    errdefer session.deinit(allocator);
    repository.save(client, config, allocator, &session) catch return error.RedisUnavailable;
    return session;
}

pub fn logout(
    client: *redis_mod.Client,
    config: *const session_mod.SessionConfig,
    allocator: std.mem.Allocator,
    session_id: []const u8,
) AuthError!void {
    repository.deleteSession(client, config, allocator, session_id) catch return error.RedisUnavailable;
}

/// Deletes the current session id (if any), writes a fresh guest session to Redis, and returns it.
pub fn logoutAndCreateGuest(
    client: *redis_mod.Client,
    config: *const session_mod.SessionConfig,
    allocator: std.mem.Allocator,
    session_id: ?[]const u8,
) AuthError!session_mod.SharedSession {
    if (session_id) |id| {
        const trimmed = std.mem.trim(u8, id, " \t\r\n");
        if (trimmed.len > 0) {
            repository.deleteSession(client, config, allocator, trimmed) catch return error.RedisUnavailable;
        }
    }
    var guest = createGuestSession(allocator) catch return error.OutOfMemory;
    errdefer guest.deinit(allocator);
    repository.save(client, config, allocator, &guest) catch return error.RedisUnavailable;
    return guest;
}

pub fn refreshSession(
    client: *redis_mod.Client,
    config: *const session_mod.SessionConfig,
    allocator: std.mem.Allocator,
    current: ?*const session_mod.SharedSession,
) AuthError!session_mod.SharedSession {
    if (current) |existing| {
        repository.deleteSession(client, config, allocator, existing.session_id) catch {};
    }
    var session = if (current) |existing| blk: {
        if (existing.user_id > 0 and existing.email != null) {
            break :blk createUserSession(allocator, existing.user_id, existing.name, existing.email.?)
                catch return error.OutOfMemory;
        }
        break :blk createGuestSession(allocator) catch return error.OutOfMemory;
    } else createGuestSession(allocator) catch return error.OutOfMemory;
    errdefer session.deinit(allocator);
    repository.save(client, config, allocator, &session) catch return error.RedisUnavailable;
    return session;
}

pub fn resolveCurrentSession(
    client: *redis_mod.Client,
    config: *const session_mod.SessionConfig,
    allocator: std.mem.Allocator,
    session_id: []const u8,
) AuthError!?session_mod.SharedSession {
    const trimmed = std.mem.trim(u8, session_id, " \t\r\n");
    if (trimmed.len == 0) return null;
    const session = repository.findById(client, config, allocator, trimmed) catch return error.RedisUnavailable;
    const resolved = session orelse return null;
    if (resolved.isExpired()) {
        var expired = resolved;
        defer expired.deinit(allocator);
        repository.deleteSession(client, config, allocator, trimmed) catch {};
        return null;
    }
    return resolved;
}

fn createGuestSession(allocator: std.mem.Allocator) !session_mod.SharedSession {
    const session_id = try session_mod.newSessionId(allocator);
    const issued_at = try session_mod.utcNowIso(allocator);
    const expires_at = try session_mod.expiresIso(allocator, session_mod.DEFAULT_TTL_SECS);
    return .{
        .session_id = session_id,
        .user_id = 0,
        .email = null,
        .name = try allocator.dupe(u8, "Guest"),
        .issued_at = issued_at,
        .expires_at = expires_at,
        .issuer = try allocator.dupe(u8, session_mod.ISSUER),
    };
}

fn createUserSession(allocator: std.mem.Allocator, user_id: i64, name: []const u8, email: []const u8) !session_mod.SharedSession {
    const session_id = try session_mod.newSessionId(allocator);
    const issued_at = try session_mod.utcNowIso(allocator);
    const expires_at = try session_mod.expiresIso(allocator, session_mod.DEFAULT_TTL_SECS);
    return .{
        .session_id = session_id,
        .user_id = user_id,
        .email = try allocator.dupe(u8, email),
        .name = try allocator.dupe(u8, name),
        .issued_at = issued_at,
        .expires_at = expires_at,
        .issuer = try allocator.dupe(u8, session_mod.ISSUER),
    };
}
